server/base/src/main/java/org/apache/accumulo/server/fs/PerTableVolumeChooser.java

```diff
@@ -18,61 +18,146 @@
+    if (log.isTraceEnabled()) {
+      log.trace("Looking up property " + TABLE_VOLUME_CHOOSER + " for Table id: " + env.getTableId());
+    }
+    final TableConfiguration tableConf = localConf.getTableConfiguration(env.getTableId());
+    String clazz = tableConf.get(TABLE_VOLUME_CHOOSER);
+
+    return createVolumeChooser(clazz, TABLE_VOLUME_CHOOSER, env.getTableId().canonicalID(), tableSpecificChooser);
+  }
+
+  private VolumeChooser getVolumeChooserForNonTable(VolumeChooserEnvironment env, ServerConfigurationFactory localConf) throws AccumuloException {
+    String property = SCOPED_VOLUME_CHOOSER(env.getScope());
+
+    if (log.isTraceEnabled()) {
+      log.trace("Looking up property: " + property);
+    }
+
+    AccumuloConfiguration systemConfiguration = localConf.getSystemConfiguration();
+    String clazz = systemConfiguration.get(property);
+    // only if the custom property is not set to we fallback to the table volume chooser setting
+    if (null == clazz) {
+      log.debug("Property not found: " + property + " using " + DEFAULT_SCOPED_VOLUME_CHOOSER);
+      property = DEFAULT_SCOPED_VOLUME_CHOOSER;
+      clazz = systemConfiguration.get(DEFAULT_SCOPED_VOLUME_CHOOSER);
+    }
+
+    return createVolumeChooser(clazz, property, env.getScope(), scopeSpecificChooser);
+  }
+
+  private VolumeChooser createVolumeChooser(String clazz, String property, String key, ConcurrentHashMap<String,VolumeChooser> cache) throws AccumuloException {
+    if (null == clazz || clazz.isEmpty()) {
+      String msg = "Property " + property + " must be set" + (null == clazz ? " " : " properly ") + "to use the " + getClass().getSimpleName();
+      log.error(msg);
+      throw new AccumuloException(msg);
+    }
+
+    VolumeChooser chooser = cache.get(key);
+    if (chooser == null || !(chooser.getClass().getName().equals(clazz))) {
+      if (log.isTraceEnabled() && chooser != null) {
+        // TODO stricter definition of when the updated property is used, ref ACCUMULO-3412
+        log.trace("Change detected for " + property + " for " + key);
+      }
+      VolumeChooser temp;
+      try {
+        temp = loadClass(clazz);
+      } catch (Exception e) {
+        String msg = "Failed to create instance for " + key + " configured to use " + clazz + " via " + property;
+        log.error(msg);
+        throw new AccumuloException(msg, e);
       }
-      final TableConfiguration tableConf = localConf.getTableConfiguration(env.getTableId());
-      chooser = tableSpecificChooser.get(env.getTableId());
       if (chooser == null) {
-        VolumeChooser temp = Property.createTableInstanceFromPropertyName(tableConf, Property.TABLE_VOLUME_CHOOSER, VolumeChooser.class, fallbackVolumeChooser);
-        chooser = tableSpecificChooser.putIfAbsent(env.getTableId(), temp);
+        chooser = cache.putIfAbsent(key, temp);
```

Comment:
OK, I am updating this section.  It appears the logic below this is not quite right anyway.